include/rmm/mr/device/tracking_resource_adaptor.hpp

```diff
@@ -0,0 +1,252 @@
+   * resource as long as it fits inside the allocation limit.
+   *
+   * The returned pointer has at least 256B alignment.
+   *
+   * @throws `rmm::bad_alloc` if the requested allocation could not be fulfilled
+   * by the upstream resource.
+   *
+   * @param bytes The size, in bytes, of the allocation
+   * @param stream Stream on which to perform the allocation
+   * @return void* Pointer to the newly allocated memory
+   */
+  void* do_allocate(std::size_t bytes, cudaStream_t stream) override
+  {
+    void* p = [&]() {
+      try {
+        return upstream_->allocate(bytes, stream);
+      } catch (std::exception const& e) {
+        RMM_LOG_ERROR("[A][Stream {}][Upstream {}B][FAILURE {}]",
+                      reinterpret_cast<void*>(stream),
+                      bytes,
+                      e.what());
+        throw;
+      }
+    }();
+
+    // track it.
+    {
+      lock_t lock(mtx);
+      allocations.emplace(p, bytes);
+    }
+
+    return p;
+  }
+
+  /**
+   * @brief Free allocation of size `bytes` pointed to by `p`
+   *
+   * @throws Nothing.
+   *
+   * @param p Pointer to be deallocated
+   * @param bytes Size of the allocation
+   * @param stream Stream on which to perform the deallocation
+   */
+  void do_deallocate(void* p, std::size_t bytes, cudaStream_t stream) override
+  {
+    {
+      lock_t lock(mtx);
+      allocations.erase(p);
+    }
+    upstream_->deallocate(p, bytes, stream);
+  }
+
+  /**
+   * @brief Compare the upstream resource to another.
+   *
+   * @throws Nothing.
+   *
+   * @param other The other resource to compare to
+   * @return true If the two resources are equivalent
+   * @return false If the two resources are not equal
+   */
+  bool do_is_equal(device_memory_resource const& other) const noexcept override
+  {
+    if (this == &other)
+      return true;
+    else {
+      tracking_resource_adaptor<Upstream> const* cast =
+        dynamic_cast<tracking_resource_adaptor<Upstream> const*>(&other);
+      if (cast != nullptr)
+        return upstream_->is_equal(*cast->get_upstream());
+      else
+        return upstream_->is_equal(other);
+    }
+  }
+
+  /**
+   * @brief Get free and available memory from upstream resource.
+   *
+   * @throws `rmm::cuda_error` if unable to retrieve memory info.
+   *
+   * @param stream Stream on which to get the mem info.
+   * @return std::pair contaiing free_size and total_size of memory
+   */
+  std::pair<std::size_t, std::size_t> do_get_mem_info(cudaStream_t stream) const override
+  {
+    return upstream_->get_mem_info(stream);
+  }
+
+  struct allocation_info {
+    std::vector<void*> stack_ptrs;
+    std::size_t allocation_size;
+
+    allocation_info() : allocation_size(0){};
+    allocation_info(std::size_t size) : allocation_size(size)
+    {
+      // store off a stack for this allocation
+      const int MaxStackDepth = 64;
+      void* stack[MaxStackDepth];
+      auto depth = backtrace(stack, MaxStackDepth);
```

Comment:
Don't you want the backtrace at the point of the call to allocate? Perhaps call backtrace in do_allocate and pass the results to the allocation_info constructor?

Reply:
Also, I think that getting backtraces for every allocation should be optional. We want to enable detecting leaks in our CI tests, but we don't want to slow down the tests every run. Instead, we should be able to report that there are leaks, and their sizes, and then we can re-run the tests with a switch flipped to enable the BT logging for each leak. (This would be a command line switch added to the tests in e.g. libcudf)

Reply:
Can you offer guidance on how to accomplish this? I'm not familiar with the testing framework. Adding a function to enable/disable stacks is easy enough, but hooking it into the test framework is my question. Would you want this on a global level or on a per-allocator basis?

Reply:
> Don't you want the backtrace at the point of the call to allocate? Perhaps call backtrace in do_allocate and pass the results to the allocation_info constructor?

I thought about this, but we're adding a single extra stack frame to each stack. I didn't see it as a big deal and would prefer to keep all that logic and even knowing about the stack trace inside the constructor, but I'm happy to move the code out if desired.